src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/DirectiveCompletionItemProvider.cs

```diff
@@ -27,6 +27,24 @@ internal class DirectiveCompletionItemProvider : RazorCompletionItemProvider
             CSharpCodeParser.TagHelperPrefixDirectiveDescriptor,
         };
 
+        // internal for testing
+        internal static readonly IReadOnlyDictionary<string, string> s_singleLineDirectiveSnippets = new Dictionary<string, string>()
```

Comment:
@DoctorKrolic super cool PR!

@mikadumont @danroth27 @DamianEdwards how do you feel about single line directive completions being output as snippets by default? Do you think we need to offer a non-snippet variant? Should we opt-towards the C# model of `tab + tab` be the invocation? Should we just go right into the snippet (what's done in this PR)?

I pulled this PR and here's how it works in practice:
![image](https://i.imgur.com/PSl9fvI.gif)
![image](https://i.imgur.com/znmLZUY.gif)

If so, what sort of content would you want on the right hand side of all these directives? Does the content @DoctorKrolic added make sense or would we want a different default?

Reply:
> Should we opt-towards the C# model of tab + tab be the invocation? 

1. As far as I know, semantic snippets will move from `tab + tab` model to a single `tab` and keep `tab + tab` just for "muscle memory backward compatibility". Well, we can teach our users to use single `tab` from the very beginning, so we won't need to do similar trick later
2. `tab + tab` is a bit more tricky to implement in the first place
3. Single `tab` feels more LSP way of doing it
4. You rejected my PR, that removes adding C# snippets via post-processing because of discoverability reasons - https://github.com/dotnet/razor-tooling/pull/6107#issuecomment-1058815363. Well, the same discoverability reason can be applied to this functionality as well
5. `tab + tab` doesn't seem to have any major advantages except consistency with current C# solution. But considering `(1)` this might not be the case in the long run

> Does the content @DoctorKrolic added make sense from a runtime perspective or would we want a different default?

As about defaults, I can split them into 3 categories. The first are those which have strong rational arguments behind and those I would promote to keep "as is":
 - `@implements IDisposable` - from my experience implementing `IDisposable` is a common practice for razor components. I don't think we loose anything by providing this interface by default, but acquire many (potentially) use cases, where user just hit `tab` to insert completion and `enter` to "agree" with the given interface. Less typing -> better productivity
 - `@preservewhitespace true` - according do docs, if this directive is not specified, than it would be implicitly `false`. So the vast majority of cases, when user is typing this directive is when he wants to switch it to `true`
 - `@typeparam T` - well, ask any C# developer including yourself and you will find out that `T` is a default type parameter name that comes into head and it is widely used in real code bases
 
 The second category contains items, that also have rational arguments behind, but I don't feel very confident about them:
 - `@inherits MyRazorPageClass`, `@inject IService MyService`, `@namespace MyNameSpace`, `@model MyModelClass` - in C# names that starts with `My` are common generated ones. For instance, executing `class`, `struct` or `record` snippets will produce `MyClass`, `MyStruct` and `MyRecord` as default names. Executing `interface` snippet produces `IInterface` default name - this is the reason I've chosen the `IService` default name for `@inject`
 
 And the third category is for items, where I picked something without strong rational arguments in mind:
 - `@add(remove)tagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers` - these values are kinda default
 - `@attribute [Attribute]` - well, you want an attribute, so I am explicit about it 🙂
 - `@page "/page"` - just an example of how a page endpoint should look like, I don't have much to add here
 - `@taghelperPrefix prefix` - I genuinely dodn't know what is better to put here
 - `@layout MainLayout` - `MainLayout` should exist in every default blazor app, so let it be here
 
 Some names from the third category can be taken to the second with the same `My` naming in mind, like `@layout MyLayout`. Maybe you will come up with something even better for the second and third categories. Again, I am confident only in the first one
 
 Uh, and the last thing: you raised so many questions that are enough for a design meating. This is kinda fun, considering that the underlying issue was not touched for 5 years. I definitely want this questions to be resolved in less than another 5-year period. While testing my work I always had a thought in mind: "Wow, that is so convenient to use! I wonder, why they haven't implemented these things earlier..."

Reply:
@DoctorKrolic love the passion you're contributing here. @danroth27 / @DamianEdwards are oof for a lot of the runtime questions until beginning of August so we'll probably need to wait on them for that bit (although I'll make sure we can have an agreed upon path in 17.4).

I loved your defaults for the directives; some areas where I could potentially see deviation is to make them even simpler to indicate what type of value is expected, aka `@inject TypeName PropertyName`. One of the reasons i'd love to explore everyones thoughts here.

> tab + tab is a bit more tricky to implement in the first place

Actually not too bad. We have an [InlineCompletionEndpoint](https://github.com/dotnet/razor-tooling/blob/24c6e5d76616602dbf82153981840c6746029bc4/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/InlineCompletion/InlineCompletionEndPoint.cs) which handles exactly this; however, today it's what powers `for`, `if` etc. tab + tab behavior. I agree with your other sentiments though of single tab being more approachable. My concerns lie in how we educate users how to interact with the snippet system. Aka cycling through options with tab / shift tab and finalizing with enter. Maybe that can be an addition to the description text?

> "Wow, that is so convenient to use! I wonder, why they haven't implemented these things earlier..."

Totally fair, and can absolutely see the value here. Interactions with a snippet-first system is probably our primary reason we've been hesitant; although for those who know how to interact with it this is incredibly powerful. Oh and as a snippet-knowing person I also thing this makes things super convenient.

Reply:
>  @danroth27 @DamianEdwards Does the content @DoctorKrolic added make sense from a runtime perspective or would we want a different default?

The suggested content seems reasonable. Instead of `@attribute [Attribute]` could we do `@attribute [Authorize]`, which is pretty common?

Reply:
Roslyn's semantic snippets are moving from the tab + tab model to a single tab so this single tab experience aligns well with that.

Reply:
@DoctorKrolic we're chatting offline about your PR, it's quite awesome and holy cow has the potential to change the Razor editing ecosystem a ton! To be super transparent about what we're chatting about internally:

1. How will folks who've never interacted with snippets before know how to operate them? (there's a surprising amount of folks who have 0 context of snippets)
2. Will folks be alarmed by the auto-injection of content even though it doesn't show that in the completion item display text?
3. Is there an alternate experience that would better fit the discoverability mechanisms of directive completions. i.e., if a snippet doesn't save all that much time, is its purpose to make things discoverable? If so, would signature help be better suited? Would something else? Orrrr are snippets still a net-win at completion time?
4. Is there modifications to the current feature to solve the above in a reasonable way to ensure we account for all developer levels?

Anyhow, wanted to update this thread so you didn't think we had forgotten about you. Should have some updates mid-next week.

Reply:
My own opinion about these questions:
1) Well, I am certanly not a UX designer, but isn't it obvious, if you see some text highlighted then it means it is editable? `Tab` and `Enter` keys that are used to operate snippets are very common for the whole VS in general. I mean, `Tab` is used to insert completions, apply AI suggestions, invoke `tab + tab` snippets with inline completion and so on. As for me if I see something new with the VS editor that I am not familiar with, these 2 keys are the first to be pressed)
2) That can be uncomfortable at first, but I think you can get used to it very quickly. Maybe there should be an option to bring completion experience back to what it used to be, but IMO this new behaviour should be enabled by default
3) No no no, the primary goal of snippets is to reduce typing. For instance, for me in order to type `"` I have to press `shift`, so typing `@page` directive and its content is a bit slow. With this change I will be able to focus more on "the actual logic", rather than on the surrounding "scaffolding".
4) Well, as I said above, I can only suggest adding an option switch for this feature, but leave it enabled. Later you will be able to determine if such option is used by the telemetry means. Because if user disables it, that he is doing it consciously and not because it was the default value. It it appears that this switch is not used you will just remove it to reduce code complexity. On the other side, if it is gonna be used frequently than you should collect feedback and improve the feature.

Reply:
Re item number 2: Yes, if these snippets insert with any commit character except Tab (or Tab+Tab) I would be very alarmed, and very angry.

Reply:
So we were able to chat a lot and here's the bit that was landed on:

1. We should have 2 completion items per-directive. One for the existing non-snippet variant and one for the new snippet variant.
2. The non-snippet variant should continue to be "prioritized" as a user types. Meaning, if they do `@inject` it should filter down to 2 items (non-snippet and snippet) but have the non-snippet one selected requiring the user to hit down-arrow to select the snippet variant. Hell maybe eventually we could build in smarts to change the prioritization model based on usage 😄 
3. The snippet variant should be extra clear (in addition to icon) that it's inserting additional content. There's a lot of ways we can do this through the display text. I don't have a greatttt catch-all answer for what to display so for simplicity we could start out using `...` suffixes. Meaning model could be something like `@model ...`. Ultimately this would help folks who have no idea what a snippet even is to understand that the completion item does a little bit more without requiring selection.
4. (optional) Tooltip for the directives could be a little more verbose on what's being inserted. I.e. they could state the snippet text without the `$` and have a little disclaimer below them indicating how to traverse the snippet. For instance `model ...`'s tooltip could be something like:
```
@model MyModelClass

Type to replace placeholders. [Tab] to navigate or [Return] to complete snippet.
```